Add upgrader 3.1 and support for upgrader chains

adecler opened the pull request:
### NOTE: Depends on 
https://github.com/BHoM/BHoM_Engine/pull/1477

   
### Issues addressed by this PR
Closes #20

<!-- Add short description of what has been fixed -->


### Test files
- for objects:
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/Versioning_Toolkit/Issue21_BackwardCompatibilityTest.gh?csf=1&e=FuoX56
- for methods:
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/Versioning_Toolkit/Issue7_MethodUpgradeTest_ForAdapters.gh?csf=1&e=mMnHaH

Same file as before and should give the same result. The difference is that the upgrades done for this files are now one level down the upgrader chain.

What you can do as a test is use the PR from the Engine (that correctly gets the assembly file version) and see that nothing gets upgraded properly. Then switch to this branch onV_TK and see that it works again.

The other test will obviously be when @IsakNaslundBh  and @kThorsager start adding they versioning for this quarter. 

Here's the result that you should get:

![image](https://user-images.githubusercontent.com/16853390/74218375-ea935500-4ce4-11ea-8b35-e7fea93d3f33.png)

kThorsager commented:
Tested this and will begin by saying that the provided file does not look like picture above.
This it the file I got (the red Push/Pull works for non-socket adapters):
![image](https://user-images.githubusercontent.com/53530468/74225747-2f9e9380-4cb3-11ea-88aa-d75c2f8b923b.png)

So for a master, master configuration it worked for all the adapter methods in Rhino 6 except execute, but it is not among the methods in the dictionary so expected?
It did however not work for Rhino 5, which might be problematic since we are currently having some [conversion problems in rhino 6](https://github.com/BHoM/Rhinoceros_Toolkit/issues/137)

For PRbranch, master it did not work, which was expected.

But for PRbranch PRBranch it did not work either...
Will try to figure out whats going on there.

adecler commented:
Sorry @kThorsager , I have updated the link to the test file

kThorsager commented:
No worries
In that file the objects are converted nicely to their new type, but unexpectedly well. It works even when I'm only on the BHoM_Engine branch.
And it does not contain any tests for method upgrades

adecler commented:
@kThorsager , let me start over. I rushed the PR just before a meeting yesterday and should have taken a bit more time to test and communicate. 

I have now provided links to two test files in the top comment. One for objects, one for methods. Both should still open correctly. With a few caveats:
- The file testing the objects, should open as in the picture on the top comment. The two orange components and the red one are expected behaviour and testing that versioning fails properly when upgrades are not provided.
- The file testing the adapter methods should upgrade fine except for the `Execute` methods that doesn't look like it was provided an upgrade for. @alelom, can you check on that ? I have added you as a reviewer as I should have from the start.

However, I always recommend you run your own tests based on your own usage of what is provided in a PR. In this case, I added you because you are planning on using the versioning toolkit yourself so I wanted to make sure this PR was providing what you need (i.e. the upgrader 3.1) without breaking anything else. You are obviously welcome to test from any angle you see fit.

Now, for the two things you have highlighted already:
- You are saying that everything is upgraded nicely even when you are using only the Engine PR ? This means that the 3.0 upgrader is still used somehow since the upgrades necessary are only available there. So either you were still somehow on the new version of the versioning toolkit (or didn't recompiled BHoM_UI) when you tested or your current version of the BHoM is still 3.0 (but I would find that unlikely unless you haven't updated the UI repos for a while). Can you try again with that in mind ? (see useful tip below for a better way to check what is going on)
- The first test that you did on the adapter methods was failing massively and was doing the same for me. I have found the source of the problem and added a commit here that should fix this. So if you try again, everything should work except the `Execute` component.


**_Useful tip:_** if you comment those two lines in the GetPipe() method of the `BHoM_Engine\Versioning_Engine\Convert\ToNewVersion.cs`, you will have a command window poping up for each upgrader with their name in the window title. You can then also attach a debugger to them to see better what is going on. 

![image](https://user-images.githubusercontent.com/16853390/74297458-e8330880-4d81-11ea-8362-65b395d273a7.png)

kThorsager commented:
Great, I got it to work properly now,
My issue was that once BHoM_Upgrader31 has been build it stays in `Versioning_Toolkit\Build\bin\` as nothing overwrites it in the old branch. So despite clearing the appdata folder, BHoM_UI put it back when built. (and I was a bit unsure of where exactly it went in there so I did not find it yesterday)

guess this will only be a problem for people going backwards, but slightly annoying while testing.

But I will continue doing tests for my specific cases

kThorsager commented:
So what I'm trying to do is changing `Geometry.IElement` to `Dimensional.IElement` (and the other similar interfaces)

The does't seem to be any framework in Versioning_toolkit for updating a interface. 
Now specifically updating a type component to the new type component is not strictly necessary (but would be nice).
The biggest issue is that every method with one of these interfaces as input will no longer be found.

So while trying to fix that I notice that when loading a document with components not linked to methods, this part:
![image](https://user-images.githubusercontent.com/53530468/74332291-56c99380-4d8d-11ea-8ad5-ec742061b140.png)
has `_t = DBNull` and no `System.Reflection.MethodBase` are pulled trough. Unsure if the methods are sent in as DBNull or just not sent in. (or maybe I have misunderstood this)

A bug:
if someone specifies a method here:
![image](https://user-images.githubusercontent.com/53530468/74332669-2b937400-4d8e-11ea-944e-9c293e7b3924.png)
without the parameter input for a overloaded method, then grasshopper will buffer forever(? a long time at least).

adecler commented:
@kThorsager ,

> The does't seem to be any framework in Versioning_toolkit for updating a interface.

You do it the same was as you do any class, you use `Converter.ToNewType`.

> The biggest issue is that every method with one of these interfaces as input will no longer be found.

Correct, this is currently only supported for objects but not methods. So the only way right now is to provide an update description for each method. This is way beyond the scope of this PR though so I would raise a separate issue for that.

> when we want to upgrade into something deprecated in the chain after the type has been removed it won't work

Very true.  I was planning to set up the necessary NuGet packages for this next week so that each upgrader depends on its own version of the BHoM instead of the current one. Given how little time I have at the moment, I am starting to wonder if it wouldn't be better to simply store the dlls locally even it it means pushing dlls to GitHub. @al-fisher, what do you think ?

> and it makes Versioning_Toolkit dependent on a lot of things

That is the point but keep in mind we only need the dlls that correspond to upgraded code during that quarter.

al-fisher commented:
>  I am starting to wonder if it wouldn't be better to simply store the dlls locally even it it means pushing dlls to GitHub. @al-fisher, what do you think ?

Yes. Think this is fine for now.
With plan to migrate to NuGets to scale in medium long term

kThorsager commented:
@adecler 
> > The does't seem to be any framework in Versioning_toolkit for updating a interface.
> 
> You do it the same was as you do any class, you use `Converter.ToNewType`.

Sorry, I should have been more clear:
![image](https://user-images.githubusercontent.com/53530468/74422566-b89a0400-4e46-11ea-9c0d-594827dac466.png)
This is what I meant. And it does work for every other part of it.
![image](https://user-images.githubusercontent.com/53530468/74422709-ef701a00-4e46-11ea-9555-a06638d5718b.png)
A small thing, but still

adecler commented:
Thanks @kThorsager .
I have raised a few more issues to better highlight the aspects of the versioning toolkit that still need immediate work. This toolkit is still very much in its infancy with features pretty much created on a need-to-have basis. I however plan to make sure you have everything you need by the end of next week at the latest. So don't hesitate to raise further issues if you see missing features that are preventing you from doing your job on the Geometry to Dimensional refactoring.
